### ISSUE TITLE
mypy too strict in enforcing generic return types when they are unions

### ISSUE THREAD

**Issue by thesketh** [python/mypy#19239]
**Bug Report**

mypy is overly strict in enforcing union return types from generic functions.

If a function both takes and returns a value of type T, which is a union of two types, and does any guarding to check which of the possible types the value is before returning it, mypy emits an "Incompatible return type" error.

**To Reproduce**

This happens with checks using `isinstance` or with `is None` e.g.:

```python
def generic_function[T: str | int](value: T) -> T:
    if isinstance(value, int):
        return value
    return value
```
https://mypy-play.net/?mypy=latest&python=3.12&gist=470d29ce79af159a1950e5f47aac2798

and

```python
def generic_function[T: str | None](value: T) -> T:
    if value is None:
        return value
    return value
```
https://mypy-play.net/?mypy=latest&python=3.12&gist=16f08fc7cc619fa2797d12d6c50e2103

**Expected Behavior**

mypy should not emit an error, because the input and output types are still correct, the guard just leads to an early return.

I suppose more specifically I would expect that guards on the input type for a generic which takes and returns the same type should also be guards on the output type.

**Actual Behavior**

```error: Incompatible return value type (got "int", expected "T")  [return-value]```
and
```error: Incompatible return value type (got "None", expected "T")  [return-value]```

**Your Environment**

- Mypy version used: 1.16.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
```toml
[tool.mypy]
strict = true
python_version = "3.12"
plugins = ['pydantic.mypy']
```
- Python version used: Python 3.12

**Comment by thesketh** [comment]
I've changed the title - the `None` case is a 1.16 regression and has been fixed on `master`, but the `int`/`isinstance` case still emits an error and I don't think it should.

**Comment by A5rocks** [comment]
I assume this is fixed by https://github.com/python/mypy/pull/19183

**Comment by thesketh** [comment]
Looks like it has been - I must have accidentally clicked "mypy latest" when I tried the `isinstance(value,  int)` case, whoops. Nice spot @A5rocks
